sklearn/utils/validation.py

```diff
@@ -322,20 +336,34 @@ def check_array(array, accept_sparse=None, dtype="numeric", order=None,
     # store whether originally we wanted numeric dtype
     dtype_numeric = dtype == "numeric"
 
-    if sp.issparse(array):
-        if dtype_numeric:
+    dtype_orig = getattr(array, "dtype", None)
+    if not hasattr(dtype_orig, 'kind'):
+        # not a data type (e.g. a column named dtype in a pandas DataFrame)
+        dtype_orig = None
+
+    if dtype_numeric:
```

Comment:
I'm not 100% sure if we want to keep this behavior. In what cases do we want to pass through integer data? Or bytes? But maybe that is for another PR? I'm not sure....
We could also just set dtype = NUMERIC_DTYPES (which would be a long list) to remove this branch at least.